Signal handling in DataLoader workers; Timeout option

ssnl opened the pull request:
1. Make DataLoader workers a bit more verbose on bus error and segfault, implemented with C++ signal handlers. Partially addresses https://github.com/pytorch/pytorch/issues/1355
2. Adding timeout option to DataLoader, implemented with SIGALRM. https://github.com/pytorch/pytorch/issues/2474 

Test plan (modified from script by @fmassa in #1595 ):
1. Segfault case:
script: 
```
class DS(object):
    def __getitem__(self, idx):
        import ctypes;ctypes.string_at(0)
        return torch.rand(1000000)
    def __len__(self):
        return 200

ds = DS()
it = torch.utils.data.DataLoader(ds, batch_size=10, num_workers=10)

for i, data in enumerate(it):
    print(i)
```
output: 
```
[~] python ds.py
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
[~] # Doesn't hang anymore
```

2. Timeout:
script:
```
import time

class DS(object):
    def __getitem__(self, idx):
        time.sleep(10)
        return torch.rand(1000000)
    def __len__(self):
        return 200

ds = DS()
it = torch.utils.data.DataLoader(ds, batch_size=10, num_workers=10, timeout=2)

for i, data in enumerate(it):
    print(i)
```
output:
```
[~] python ds.py
ERROR: Time out when fetching data in DataLoader.
[~]
```

Since this is deep into multiprocessing and low level signal handling, there might be cases/things I haven't considered. I'll also leave some comment below.

cc @apaszke @colesbury

ssnl committed:
signal handling on data loader workers; time out

SsnL commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+
+  auto size = tuple ? PyTuple_GET_SIZE(child_pids) : PyList_GET_SIZE(child_pids);
+  {
+    std::unique_lock<std::mutex> lock(worker_pid_mutex);
```
Is guarding this with a lock necessary? Is GIL acquired here?

apaszke replied:
You always have the GIL unless you give it up

SsnL commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+  // The flags and status checks ensure that we are really observing a child
+  // exiting, rather than other cases such as SIGSTOP and SIGCONT.
+  // https://stackoverflow.com/a/40707100
+  while ((p = waitpid(-1, &status, WNOHANG|WUNTRACED|WCONTINUED)) > 0) {
```
This will fail on windows, right?...

apaszke replied:
Most of this patch is likely to fail on windows I think. I don't think there's a similar signalling mechanism there. @peterjc123 do you have any hints for making this portable?

apaszke replied:
I'm also a bit worried of this use of `waitpid` and a custom signal handler for SIGCHLD. It might e.g. overwrite Python's handlers if it uses them to get return codes in `multiprocessing`. Also, if you use `waitpid` without `NOWAIT`, it will reap the child, so Python won't be able to retrieve the exit code.

peterjc123 replied:
A reference from the implementation from Apache [here](https://reviews.apache.org/r/46340/diff/2#index_header). They just defined a Windows version of `os.waitpid`, supporting the option arg `WNOHANG`. But i don't think there is sth like that in Windows. A better way to implement it may be using threads or processes like what i stated below.

SsnL commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+static pid_t *worker_pids;
+static size_t num_worker_pids = 0;
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
```
Let me know if I should move these methods to a separate file.

colesbury replied:
Yeah, I think it'd be better in a separate file. In the new file's header you can export a null-terminated `PyMethodDef[]`. In Module.cpp use THPUtils_addPyMethodDefs to add it to the methods vector.

colesbury commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig)                                             \
+{                                                                             \
+    write(fileno(stderr), ERROR_MSG, strlen(ERROR_MSG));                      \
```
I don't think `fileno` is async-signal-safe:

http://man7.org/linux/man-pages/man7/signal-safety.7.html

SsnL replied:
Ah... I checked most methods except fileno :P. Good catch!

apaszke replied:
`strlen` isn't officially async-signal-safe either. If `ERROR_MSG` is going to be a literal, you can define a char array in this function and just use `sizeof` to get the length

colesbury commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+// We don't want to exit on any SIGCHLD from any child. child_pids is a sequence
+// of pids we are interested in.
+PyObject *THPModule_setMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  auto tuple = PyTuple_Check(child_pids);
```
Since this is only called internally, stick to just tuples or just lists. It'll simplify the code.

colesbury replied:
Use `HANDLE_TH_ERRORS` `END_HANDLE_TH_ERRORS` in your Python C functions. They surround everything with a try/catch that turns C++ exceptions to Python exceptions.  It's safe even if you don't think things throw and it doesn't add runtime overhead.

colesbury commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+    pid_ptr = worker_pids;
+    for (size_t i = 0; i < num_worker_pids; i++) {
+      if (*pid_ptr == p)
+        _exit(EXIT_FAILURE);
```
Does this exit the main-process without an error message?

SsnL replied:
Yes. Printing something will cause the annoying behavior when the main proc errors, it kills children, then this handler gets called and prints "Worker exited unexpectedly in DataLoader", which doesn't make much sense. In such case, the main proc errors, something should already be printed onto stderr.
In case where worker died, we should get error message / segfault / bus error printing from worker. 
So there should always be something on stderr.

colesbury reviewed (commented):
This looks pretty good. Can you add an automated test:

1) That the main process exits if the child segfaults
2) That the main process exits if the get times out (use a very small timeout)

You can probably trigger a segfault in the child with ctypes.

colesbury commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+    num_worker_pids = worker_pid_vec.size();
+  }
+
+  signal(SIGCHLD, handler_SIGCHLD_main);
```
Does this ensure that the signal is only handled by the main thread?

SsnL replied:
Good question. Python's `signal` module handler always ensures main thread gets the signal. Unfortunately, on c++ in POSIX we don't have such guarantee. I can add some signal blocking to ensure that threads spawned after this won' t handle the signal, but I'm not sure what I we can do with threads spawned before this. I'll investigate more tomorrow.

apaszke replied:
Does it really matter which thread handles the signal? `_exit` should abort the whole program, right?

SsnL replied:
@apaszke is right. Somehow I mistakenly thought each thread has its separate handler.

vadimkantorov commented:
Does it address the possible hangs of work_queue.put or for work_queue.get? (haven't checked the variable names). Usually it's work_queue.get that hangs, but in some corner cases (like priming) work_queue.put can hang too (in the case all worker threads are waiting on something, say data_queue.put).

SsnL commented:
@vadimkantorov It addresses it in two cases:
1. In case where workers die and causing loader proc to hang, now the loader proc will exit since it checks in SIGCHLD handler. In addition, workers dying from segfault and bus error (e.g. shm issue) now prints some message.
2. In case where workers hang, well, there is not much we can do from PyTorch side. But the new timeout option should allow the loader to at least not hang, if you use it.

vadimkantorov commented:
@SsnL Regarding (2): There seem at least two places where the loader can hang: `data_queue.get` and `index_queue.put` (less obvious). Does the timeout apply to both?

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+    signal(SIGBUS, handler_SIGBUS);
+    signal(SIGSEGV, handler_SIGSEGV);
```
Check return values for error

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+      continue;
+    // child must have exited with signal/error, check if it is one of the pid
+    // we care about.
+    pid_ptr = worker_pids;
```
This still seems unsafe - what if the signal gets triggered in the middle of `pid_ptr` assignment? Both these values should be `std::atomic` (and probably we should assert that `is_lock_free` is true, or we might get a deadlock)

SsnL replied:
`pid_ptr` is a local variable in the handler. I don't think that it should trigger a signal.

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+    std::unique_lock<std::mutex> lock(worker_pid_mutex);
+    worker_pid_vec.erase(std::remove_if(worker_pid_vec.begin(), worker_pid_vec.end(),
+      [&](pid_t p){return pid_set.find(p) != pid_set.end();}), worker_pid_vec.end());
+    worker_pids = &worker_pid_vec[0];
```
I'm not sure how to fix this, but all this manipulation seems really unsafe. You're fiddling with the vector (making it smaller), even though `num_worker_pids` still contains the large value, so these elements at the end might still be read by the signal handler.

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+  }
+
+  signal(SIGCHLD, SIG_DFL);
+  signal(SIGALRM, SIG_DFL);
```
This assumes that you have at most one `DataLoaderIter` alive at any point, which might not be true. Why not always keep those handlers?

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -174,6 +190,15 @@ def __init__(self, loader):
+        if self.timeout > 0:
+            signal.alarm(self.timeout)
+            result = self.data_queue.get()
+            signal.alarm(0)
```
I don't really like `SIGALRM`, because it's a global thing. We might run into problems when we'll have multiple `DataLoaders` in a single process

apaszke reviewed (commented):
I'm concerned about the safety and interactions of this patch. Signals are a process-wide mechanism and might interact both with Python runtime, and mix if you have multiple data loaders. Handling `SIGCHLD` makes sense (if only it doesn't break `multiprocessing`), but I'd rather not use `SIGALRM`

peterjc123 commented:
There're two kinds of possible solutions for cross-platform SIGALARM.
1. Use a thread based model.(threading/Event) Kills/Locks the thread after timeout. Light but will not be safe if there's something to do with GIL. examples: [1](https://eli.thegreenplace.net/2011/08/22/how-not-to-set-a-timeout-on-a-computation-in-python) and [2](https://stackoverflow.com/questions/8420422/python-windows-equivalent-of-sigalrm)
2. Use a process based model.(multiprocessing/subprocess) Kills the process after timeout. It's safe and easy to implement, but will be heavy. examples: [1](https://stackoverflow.com/questions/1191374/using-module-subprocess-with-timeout) and [2](https://penandpants.com/2012/09/15/function-timeout-in-python/)

SsnL commented:
@vadimkantorov Sorry I missed your comment earlier. Why should `index_queue.put()` hang? According to my understanding, only the main proc puts to that queue.

vadimkantorov commented:
@SsnL It will lock if DataLoader puts more than 64Kb (SimpleQueue is implemented by default using system pipes). This may happen in the extreme case when index is some serialized object or a very long array.

Then, if for some reason the reason the workers are blocked too, no-one will read from the other side of `index_queue`, so `index_queue.put` would lock indefinitely. This may theoretically happen during data loader priming, when the main thread can't consume from `data_queue`, so in the case of small number of workers and large serialized batch objects and indices data loader priming should hang.

I tried to showcase this scenario here but hit another problem first: https://github.com/pytorch/pytorch/issues/2474#issuecomment-323203974

SsnL commented:
@vadimkantorov Thanks for the explanation! Unfortunately this PR won't solve that issue. Hopefully such an extreme issue won't come up often :)

vadimkantorov commented:
@SsnL This would happen even if indices object is small, but the workers for some reason are very slow or locked (it just would take some time to fill 64Kb queue). Can we detect if the queue contains a near-critical number of bytes before putting on it? and showing a UserWarning the first time it happens

Though it would sure complicate the logic.

ssnl committed:
timeour: SIGALRM -> queue.Queue; cpp code in new file; fix some safety concerns; tests

SsnL commented:
@vadimkantorov I see. But that is not easy to do in this PR, especially now that I moved away from using SIGALRM due to concerns raised by @apaszke . I think adding another thread to data loader will solve it, but I'm not sure if it is necessary. :) I'll keep this in mind, thanks for bringing it up!

SsnL commented:
Addressing the comments by @colesbury and @apaszke , I updated PR:

1. Removed SIGALRM logic. Now the DataLoader will create a separate thread if **either**`pin_memory` is set or `timeout > 0`. Since the thread communicates with the main thread with a `queue.Queue`, we do timeout there. 

2. The SIGCHLD handler now only `waitpid` on the child processes of interest, so Python can update the `is_alive()` and `exitcode` for other processes. 

3. Re-wrote the part to remove pids from the collections so that at any time, the array contains only valid pids (some might be dead workers), and includes all alive pids.

4. Added `std::atomic` on several variables. Did `is_log_free` check.

5. Moved the new C++ functions to a separate file. Added dummy implementation for Windows.

6. Add tests. Unfortunately, for `test_segfault`, the print to `stderr` is from C++, so I don't know if there is a way to hide that on Python side. So passing the tests will still show a single line error message `ERROR: Unexpected segmentation fault encountered in worker.`.

SsnL commented on torch/csrc/jit/ir.cpp:
```diff
@@ -41,6 +41,7 @@ void printNodeRef(std::ostream & out, const Node * n) {
 template <typename T>
 std::ostream& operator<<(std::ostream & out, const std::vector<T> & nodes) {
   out << at::ArrayRef<T>{nodes};
+  return out;
```
This causes a compile time warning. So I fixed it as well.

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -174,6 +190,17 @@ def __init__(self, loader):
+                return self.data_queue.get(True, self.timeout)
+            except queue.Empty:
+                ex = RuntimeError('DataLoader timed out after {} seconds'.format(self.timeout))
+                ex.__cause__ = ex  # prevent exception chaining in python 3
```
Is this an acceptable way to prevent showing exception chain?

apaszke replied:
No idea, don't know how this works in Py2. Maybe let's just ignore it and show the chain?

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,191 @@
+  auto size = PyTuple_GET_SIZE(child_pids);
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = PyTuple_GET_ITEM(child_pids, idx);
+    worker_pid_vec.push_back((pid_t) THPUtils_unpackLong(obj));
```
I removed the lock here since, if I understand correctly, GIL is always held unless I give it up. Please correct me if I am wrong :)

vadimkantorov commented:
@SsnL I saw you used queue.Queue (instead of multiprocessing.SimpleQueue for data_queue if timeout > 0: https://github.com/SsnL/pytorch/blob/55158acd5c8c51e6ebad13f52a53c108580d8caf/torch/utils/data/dataloader.py#L176-L177

Could you use queue.Queue for index_queue in that case as well? put method also has an optional timeout argument

And probably queue.Queue has a larger message size compared to multiprocessing.SimpleQueue.

SsnL commented:
Yes I used that. But Queue is already there for interthread communication.
I just extended its usage. The inter process communication is still via
SimpleQueue. Adding one for index_queue would mean adding another thread
just for timeout. I’m not sure that is what we want. Maybe @colesbury or
@apaszke has more thoughts on this?

On Mon, Nov 6, 2017 at 19:42 Vadim Kantorov <notifications@github.com>
wrote:

> @SsnL <https://github.com/ssnl> I saw you used queue.Queue (instead of
> multiprocessing.SimpleQueue for data_queue if timeout > 0:
> https://github.com/SsnL/pytorch/blob/55158acd5c8c51e6ebad13f52a53c108580d8caf/torch/utils/data/dataloader.py#L176-L177
>
> Could you use queue.Queue for index_queue in that case as well? put method
> also has an optional timeout argument
>
> And probably queue.Queue has a larger message size compared to
> multiprocessing.SimpleQueue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/3474#issuecomment-342336884>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFaWZZriAcCMSiL_yvJ5DuoxE1AHvIqaks5sz6dngaJpZM4QRtCP>
> .
>

colesbury commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,191 @@
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
```
Does Python install it's own signal handlers? This doesn't look like it calls the previously installed signal handler.

SsnL replied:
No python doesn't.

colesbury replied:
There are a few hits for SIGCHLD in the Python source code that look relevant:

https://github.com/python/cpython/blob/fc6b348b12ad401cab0261b7b71a65c60a08c0a8/Lib/multiprocessing/forkserver.py#L173

https://github.com/python/cpython/blob/3e2ad8ec61a322370a6fbdfb2209cf74546f5e08/Lib/asyncio/unix_events.py#L793

https://github.com/python/cpython/blob/03eb11f0b354751248b427455b89e9340cfd2b47/Modules/posixmodule.c#L5926

SsnL replied:
Some indeed look relevant :) Thanks, I'll make sure to call the original handler.

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,191 @@
+
+// signal(2) is really not portable. So use sigaction.
+// http://man7.org/linux/man-pages/man2/signal.2.html
+#define SET_SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_FLAG)                  \
```
I wanted to make the code cleaner.. But I can spell it out.

colesbury replied:
I meant: can you use a function instead of a macro here?

apaszke replied:
Why would that exception end a process? It's just a helper for setting signal handlers. It's run in one of the regular threads and the exception will be caught and reported in Python

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -30,6 +34,11 @@ def _worker_loop(dataset, index_queue, data_queue, collate_fn):
 
+    # Intialize C side signal handlers because Python signal handler on things
+    # like SIGBUS and SIGSEGV will still be executed in C and likely cause error
+    # again.
```
I don't think this message is correct. Python signal handlers will cause the error again exactly because the handling will be done on the Python side and not on the C side (the C side only registers that the signal happened and returns)

SsnL replied:
This is a reworded version of Python doc 18.8.1.1

apaszke replied:
But it doesn't match the docs. They say this (emphasis mine):

> **A Python signal handler does not get executed inside the low-level (C) signal handler**. Instead, the low-level signal handler sets a flag which tells the virtual machine to execute the corresponding Python signal handler at a later point

SsnL replied:
Good point. I'll update this comment.

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -45,7 +54,7 @@ def _worker_loop(dataset, index_queue, data_queue, collate_fn):
 
 
-def _pin_memory_loop(in_queue, out_queue, done_event):
+def _process_loop(in_queue, out_queue, done_event, pin_memory):
```
`_process_loop` doesn't really tell you what it does. Maybe `_worker_manager_loop`?

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -234,11 +262,23 @@ def _shutdown_workers(self):
             self.shutdown = True
             self.done_event.set()
+            # if process_thread is waiting to put
+            if not self.data_queue.empty():
```
SimpleQueue doesn't have an `empty()` method

SsnL replied:
I'm pretty sure that it does... https://docs.python.org/2/library/multiprocessing.html#multiprocessing.multiprocessing.queues.SimpleQueue.empty

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -234,11 +262,23 @@ def _shutdown_workers(self):
             for _ in self.workers:
                 self.index_queue.put(None)
+            # if all workers hang, still put None to let process_thread exit
+            if self.worker_result_queue.empty():
```
As above. How would sending `None` to workers help with `process_thread`?

SsnL replied:
This is meant to send the exit signal to the `process_thread`. The original code relies on workers to put `None` after receiving `None` so that `process_thread` can exit.

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -268,17 +308,25 @@ class DataLoader(object):
         self.drop_last = drop_last
 
+        if not isinstance(timeout, int):
+            raise ValueError('timeout must be an integer')
```
I don't think this is necessary. We usually don't do type checks, and it would be fine for the timeout to be a float or long. Just check that it's `>= 0` if you want

SsnL replied:
I forgot to remove this :) It was needed when using SIGALRM. Thanks for pointing out.

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,191 @@
+      write_it++;
+      remaining_num++;
+    }
+  }
```
I don't know how `remove_if` is implemented. But this should make sure that the array contents are valid..., unless `resize` does something weird. I'll have to check that out.

apaszke replied:
`remove_if` will move all elements that should be deleted to the end of the container. When you call `resize` later it will drop them

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,191 @@
+
+  // Need to restore original handler so that in case DataLoader errors, the
+  // waitpids in hander won't block Python from updating Process.is_alive() and
+  // Process.exitcode, etc.
```
Yeah. I found out when I call `waitpid`, on Python side I won't see `child_proc.is_alive()` and `child_proc.exitcode` set correctly.

ssnl committed:
Addressed comments

SsnL commented:
Updated PR for @colesbury 's and @apaszke 's comments:

1. Use `WNOWAIT` with `waitid` so that Python can get the status.

2. Call the overridden SIGCHLD handler.

3. Changed the set handler macro to a helper function.

4. Now the pids array and vector do not share same underlying storage. Instead, a helper function is added to copy vector content to the array.

5. Hide stderr outputs in tests by `sys.stderr.close()`

6. Fixed outdated comments and checks. Addressed other comments on best practices.

colesbury commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,204 @@
+    worker_pids = new_ptr;
+    num_worker_pids = new_size;
+  }
+  free(old_ptr);
```
I still don't see how this is supposed to be thread-safe or signal-safe. The signal handler could be called while this function is executing or it could be using the old_ptr which is freed here.

SsnL replied:
If the handler is called before `free`, shouldn't it be using `worker_pids` instead? I am under the impression that signal handler can be called during this function but will suspend this function while executing. My thought is that this is thread-safe as it is guarded by GIL. Maybe I'm wrong..

colesbury replied:
It may not be the same thread. I don't see any code that blocks the signal handler for other threads. Even if you did that, there's no guarantee that THPModule_setMainSignalHandlers is only called from a single thread.

apaszke replied:
Well the signal can be handled in a thread that's different than the one running this function. Signal handlers can be parallel with parts of the program

SsnL replied:
Thanks, I added mechanism to prevent update and handler to run in parallel. :)

ssnl committed:
Add thread-safe signal-safe mechanics to updating pids array

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,253 @@
+    new_ptr[idx] = worker_pid_vec[idx];
+  }
+
+  // Block SIGCHLD handler for this thread so SIGCHLD handler can't happend
```
Okay, here is what happens. When we update the array, the updating thread:

1. blocks SIGCHLD signal from this thread, so the handler can't interrupt the update on this thread.
2. reads from a pipe, which acts like a lock, and writes to it after finishing. (The handler similarly also reads from the lock before reading the contents of the array, and write afterwards. )

Since multiple threads can't run this function in parallel due to GIL, I think this should work.

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,253 @@
+    new_ptr[idx] = worker_pid_vec[idx];
+  }
+
+  // Block SIGCHLD handler for this thread so SIGCHLD handler can't interrupt
```
Okay, here is what happens. When we update the array, the updating thread:

1. blocks SIGCHLD signal from this thread, so the handler can't interrupt the update on this thread.
2. reads from a pipe, which acts like a lock, and writes to it after finishing. (The handler similarly also reads from the lock before reading the contents of the array, and write afterwards. )

Since multiple threads can't run this function in parallel due to GIL, I think this should work.

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,253 @@
+}
+
+// returns -1 on error.
+static int isSIGCHLDHanderSet() {
```
I really thing we shouldn't follow the C error checking conventions in our C++ code. It's unnecessarily complicated and exceptions make these things a lot easier. Just raise an error if an error happens, and return a boolean value otherwise

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,253 @@
+  }
+}
+
+#define SAFECALL(CALL) if ((CALL) == -1) Py_RETURN_FALSE
```
That's really weird. It's not `SAFECALL` it's `IGNORECALL`. You never report errors

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,253 @@
+  }
+  worker_pid_vec.erase(std::remove_if(worker_pid_vec.begin(), worker_pid_vec.end(),
+     [&](pid_t p){return pid_set.find(p) != pid_set.end();}), worker_pid_vec.end());
+  updatePIDsArray();
```
That's why we should use exceptions. You forgot to check the return code

apaszke replied:
Also, since we're going to reallocate the pid pointer anyway, why don't we just keep a `pid_set` instead of that vector?

ssnl committed:
change to errors

SsnL commented:
Addressed @apaszke 's comments above!

pietern commented:
@pytorchbot retest this please

pietern commented:
@pytorchbot test this please

pietern commented:
@pytorchbot test this please

SsnL commented:
Requesting review on new updates. @colesbury @apaszke Thanks! :)

SsnL commented:
If there are safety concerns about SIGCHLD handler, I can remove all those and only leave the worker SIGSEGV and SIGBUS handlers.

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,244 @@
+  }
+  // Acquire ``lock'' so handlers on other threads can't interrupt
+  char c;
+  read(comm_pipe[0], &c, 1);
```
It would be nice to encapsulate this pipe in a C++ struct that conforms to the mutex protocol, so we could actually use `lock_guard`s instead of those calls to `read` and `write`

apaszke replied:
BTW I'm pretty sure this this locking is potentially susceptible to deadlocks - if your process is limited to a single core, the signal handler might interrupt this function while it holds the "lock", but it won't be able to acquire it, and will wait forever. In general, I think the only sane solution we could use is `std::atomic` for the pointer and size.

SsnL replied:
Good point. I'm already using `std:atomic`, but I think I still need to think of how to properly free old array pointer.

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,244 @@
+  // To avoid chain calling our handler, check if the current handler is already
+  // set as ours.
+  if (!isSIGCHLDHanderSet()) {
+    setSignalHandler(SIGCHLD, &handler_SIGCHLD_main, &orig_SIGCHLD_sa);
```
We should have a global lock for adding / removing workers, or we can race with setting / removing handlers

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -242,11 +269,23 @@ def _shutdown_workers(self):
+            # if all workers hang, no None is showed to worker_manager_thread,
+            # we put None to let worker_manager_thread exit
+            if self.worker_result_queue.empty():
+                self.worker_result_queue.put(None)
```
Can't we just always put it to be sure? Having a non-empty queue doesn't imply that there is a worker that didn't hang - they might freeze after queueing some results

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -242,11 +269,23 @@ def _shutdown_workers(self):
+
+    def _remove_handers(self):
+        if self.handlers_set:
+            self.handlers_set = not _remove_main_signal_handlers_for_workers(self.worker_pids)
```
Can you just make the method return nothing, and always set the value to `False`? It's not immediately clear what the return value is supposed to be (as far as I see it's a `True` constant)

ssnl committed:
use python side SIGCHLD

SsnL commented:
After thinking about this, I realize that we can safely use Python's signal module to register a pure Python side SIGCHLD handler because SIGCHLD is not an error and thusnot executing in C is not an issue. 

In the Python SIGCHLD handler, I call a C side function which then access a map of id(DataLoader)->set_of_worker_pids. The function checks each worker of each loader, and throws a runtime error if any of them errored.

SsnL commented on torch/csrc/DataLoader.cpp:
```diff
@@ -0,0 +1,166 @@
+          (infop.si_code == CLD_DUMPED)) {
+        std::ostringstream oss;
+        oss << "DataLoader worker (pid " << pid << ") exited unexpectedly.";
+        pid_set.clear();
```
This is necessary. Otherwise, the runtime error will kill the other workers, and trigger this again.

apaszke commented:
If you're going to move the handler to Python, why can't you move all the data structures to Python too?

SsnL commented:
@apaszke Because `waitid` is not available in Python 2.7

SsnL commented:
Oh no these gloo changes again D: How did you solve these last time @soumith ?

ssnl committed:
lint

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -134,6 +144,23 @@ def pin_memory_batch(batch):
+    try:
+        signal.signal(signal.SIGCHLD, handler)
+    except ValueError as _:
+        return  # Windows doesn't support this
```
Can you just make it a no-op on windows? Ignoring all ValueErrors seems to be too much

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -134,6 +144,23 @@ def pin_memory_batch(batch):
+    def handler(signum, frame):
+        _error_if_any_worker_fails()
+        if callable(previous_handler):
+            previous_handler(signum, frame)
```
Can you do this `callable` check only once outside of this function and set it to `None` otherwise? It's not particularly cheap

apaszke replied:
BTW did you verify you can still get exit codes etc. in Python (i.e. we do not interfere with the usual handlers)

SsnL replied:
I just verified and Python can indeed still get the status and exit codes etc.

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -16,6 +19,7 @@
 
 
 _use_shared_memory = False
+_SIGCHLD_handler_set = False
```
nit: can you move the global right above the function that should run once?

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -242,11 +286,25 @@ def _shutdown_workers(self):
+            # if all workers hang, no None is sent to worker_manager_thread, we
+            # put None to let worker_manager_thread exit
+            # empty check prevents put from hanging
+            if self.worker_result_queue.empty():
```
I think I mentioned in one of the previous PRs that having non-empty queue doesn't imply that there's None in it. I think it's ok to just add it unconditionally (maybe check that it's not full to avoid blocking)

SsnL replied:
Yes, I remember the comment. I didn't remove it exactly because of worrying about blocking. I'll see if I can check if it is full.

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -242,11 +286,25 @@ def _shutdown_workers(self):
             self.done_event.set()
+            # if worker_manager_thread is waiting to put
+            if not self.data_queue.empty():
+                self.data_queue.get()
```
One `get()` might not be enough. Do this in a loop

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -242,11 +286,25 @@ def _shutdown_workers(self):
 
     def __del__(self):
         if self.num_workers > 0:
+            self._remove_worker_pids_information()
```
Can you make it part of `_shutdown_workers()` the ordering is important and it's easy to forget to call this one every time. You can just inline it (the function has 3 lines)

apaszke reviewed (commented):
That looks much better. I have a few minor comments, but should be good to go after that

ssnl committed:
address @apaszke 's comments

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -45,7 +44,6 @@ def _worker_loop(dataset, index_queue, data_queue, collate_fn):
         r = index_queue.get()
         if r is None:
-            data_queue.put(None)
             break
```
For maneger_thread, `self.done_event' and manually sending `None` from main thread is more reliable than workers.

apaszke reviewed (approved):
Looks good. The only thing I thought about now is that I think `WNOWAIT` is a Linux-only thing. It would be good to verify that this diff doesn't break OS X builds.

ssnl committed:
fix set clear to use pointer

SsnL commented:
The set was not properly clearly. This is fixed in new commit. 

@apaszke I tested on OSX and they all work just as in CentOS except that the main process can't get what signal kills the worker process. But I don't think it is a big issue. Here is an output comparison:

OSX:
```
ERROR: Unexpected segmentation fault encountered in worker.
Traceback (most recent call last):
  File "ds.py", line 17, in <module>
    for i, data in enumerate(it):
  File "/Users/ssnl/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 254, in __next__
    idx, batch = self._get_batch()
  File "/Users/ssnl/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 233, in _get_batch
    return self.data_queue.get()
  File "/Users/ssnl/anaconda3/lib/python3.6/multiprocessing/queues.py", line 342, in get
    res = self._reader.recv_bytes()
  File "/Users/ssnl/anaconda3/lib/python3.6/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/Users/ssnl/anaconda3/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/Users/ssnl/anaconda3/lib/python3.6/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
  File "/Users/ssnl/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 163, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 16545) is killed by signal: Unknown signal: 0.
```

CentOS
```
ERROR: Unexpected segmentation fault encountered in worker.
Traceback (most recent call last):
  File "ds.py", line 17, in <module>
    for i, data in enumerate(it):
  File "/home/ssnl/sftp/pytorch/torch/utils/data/dataloader.py", line 254, in __next__
    idx, batch = self._get_batch()
  File "/home/ssnl/sftp/pytorch/torch/utils/data/dataloader.py", line 233, in _get_batch
    return self.data_queue.get()
  File "/home/ssnl/anaconda3/lib/python3.6/multiprocessing/queues.py", line 343, in get
    res = self._reader.recv_bytes()
  File "/home/ssnl/anaconda3/lib/python3.6/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/home/ssnl/anaconda3/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/home/ssnl/anaconda3/lib/python3.6/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
  File "/home/ssnl/sftp/pytorch/torch/utils/data/dataloader.py", line 163, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 1864508) is killed by signal: Segmentation fault.
```

ssnl committed:
use sigaction in worker handler

SsnL commented:
Sorry I forgot error checking for reset signal handler in worker handler. I also shouldn't use the nonportable signal(2). Fixed in new commit.

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -28,8 +28,14 @@
+  if (sigemptyset(&sa.sa_mask) != 0 || sigaction(SIGNAL, &sa, NULL) != 0) {   \
+    _exit(EXIT_FAILURE);                                                      \
+  } else {                                                                    \
+    raise(SIGNAL);                                                            \
```
Shouldn't you get the signal code from `info`? I didn't even know `SIGNAL` existed

SsnL replied:
`SIGNAL` is an argument of the macro. Both `sig` and `info` contain the same signal code, but those are generated at runtime when signal happens. So all three things are equivalent.

apaszke commented on torch/csrc/DataLoader.cpp:
```diff
@@ -38,9 +44,8 @@ static void setSignalHandler(int signal, void(*handler)(int, siginfo_t *, void *
-  sa.sa_flags = SA_RESTART|SA_SIGINFO|SA_NOCLDSTOP;
-  sigemptyset(&sa.sa_mask);
-  if (sigaction(signal, &sa, old_sa_ptr) != 0) {
+  sa.sa_flags = SA_RESTART|SA_SIGINFO|SA_NOCLDSTOP|SA_NODEFER;
```
what's the new flag?

SsnL replied:
It allows the same signal to be triggered within the handler rather than being deferred after handler (http://man7.org/linux/man-pages/man2/sigaction.2.html).

Actually, upon thinking about it. It doesn't do anything for this, because raising the signal is the last thing that handler do. Should I remove it?

apaszke replied:
Nah, that should be fine. Thanks for the description!
